registration/include/pcl/registration/impl/ndt.hpp

```diff
@@ -385,64 +358,51 @@ NormalDistributionsTransform<PointSource, PointTarget>::updateDerivatives (Eigen
-        hessian (i, j) += e_x_cov_x * (-gauss_d2_ * x_trans.dot (cov_dxd_pi) * x_trans.dot (c_inv * point_gradient_.col (j)) +
+        hessian (i, j) += e_x_cov_x * (-gauss_d2_ * x_trans.dot (cov_dxd_pi) * x_trans.dot (c_inv * point_jacobian_.col (j)) +
                                     x_trans.dot (c_inv * point_hessian_.block<3, 1>(3 * i, j)) +
-                                    point_gradient_.col (j).dot (cov_dxd_pi) );
+                                    point_jacobian_.col (j).dot (cov_dxd_pi) );
       }
     }
   }
 
-  return (score_inc);
+  return score_inc;
 }
 
 
 template<typename PointSource, typename PointTarget> void
 NormalDistributionsTransform<PointSource, PointTarget>::computeHessian (Eigen::Matrix<double, 6, 6> &hessian,
-                                                                        PointCloudSource &trans_cloud, Eigen::Matrix<double, 6, 1> &)
+                                                                        PointCloudSource &trans_cloud)
 {
-  // Original Point and Transformed Point
-  PointSource x_pt, x_trans_pt;
-  // Original Point and Transformed Point (for math)
-  Eigen::Vector3d x, x_trans;
-  // Occupied Voxel
-  TargetGridLeafConstPtr cell;
-  // Inverse Covariance of Occupied Voxel
-  Eigen::Matrix3d c_inv;
-
   hessian.setZero ();
 
   // Precompute Angular Derivatives unessisary because only used after regular derivative calculation
-
   // Update hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
-  for (std::size_t idx = 0; idx < input_->points.size (); idx++)
+  for (std::size_t idx = 0; idx < input_->size (); idx++)
   {
-    x_trans_pt = trans_cloud[idx];
+    // Transformed Point
+    const auto& x_trans_pt = trans_cloud[idx];
 
     // Find nieghbors (Radius search has been experimentally faster than direct neighbor checking.
     std::vector<TargetGridLeafConstPtr> neighborhood;
     std::vector<float> distances;
     target_cells_.radiusSearch (x_trans_pt, resolution_, neighborhood, distances);
 
-    for (typename std::vector<TargetGridLeafConstPtr>::iterator neighborhood_it = neighborhood.begin (); neighborhood_it != neighborhood.end (); ++neighborhood_it)
+    for (const auto& cell: neighborhood)
     {
-      cell = *neighborhood_it;
-
-      {
-        x_pt = (*input_)[idx];
-        x = Eigen::Vector3d (x_pt.x, x_pt.y, x_pt.z);
-
-        x_trans = Eigen::Vector3d (x_trans_pt.x, x_trans_pt.y, x_trans_pt.z);
+      // Original Point
+      const auto& x_pt = (*input_)[idx];
+      Eigen::Vector3d x = x_pt.getVector3fMap().template cast<double>();
 
-        // Denorm point, x_k' in Equations 6.12 and 6.13 [Magnusson 2009]
-        x_trans -= cell->getMean ();
-        // Uses precomputed covariance for speed.
-        c_inv = cell->getInverseCov ();
+      // Denorm point, x_k' in Equations 6.12 and 6.13 [Magnusson 2009]
+      Eigen::Vector3d x_trans = x_trans_pt.getVector3fMap().template cast<double>() - cell->getMean ();
```

Comment:
Both `x_trans` and `c_inv` can be made const (if `updateHessian` is also changed). Same with `computePointDerivatives`

That'd be an ABI break, but we're already doing that here, as per the label

Reply:
I added const to all unchanged variables I found in the code. I also added const qualifiers to some functions.